### PR TITLE
Cloud Bigtable: instanceAdmin sample

### PR DESCRIPTION
**Pull request by elisheva-qlogic**

### CODE REVIEW

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An example of using Google Cloud Bigtable.
+ *
+ * <p>This example demonstrates the usage of BigtableInstanceAdminClient to create, configure, and
+ * delete Cloud Bigtable Instances and Clusters.
+ *
+ * <pre>
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: might be better to use an ```<ul>```  and ```<li>```

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+  public InstanceAdmin(String projectId, String instanceId, String clusterId) throws IOException {
+    this.instanceId = instanceId;
+    this.clusterId = clusterId;
+
+    // [START connecting_to_bigtable]
+    // Creates the settings to configure a bigtable instance admin client.
+    BigtableInstanceAdminSettings instanceAdminSettings =
+        BigtableInstanceAdminSettings.newBuilder()
+            .setProjectName(ProjectName.of(projectId))
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: Please migrate away from using typesafe names and use the raw project id

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+              .addCluster(clusterId, "us-central1-f", 3, StorageType.SSD)
+              .setType(Type.PRODUCTION)
+              .addLabel("example", "instance_admin");
+      // Creates a production instance with the given request.
+      try {
+        Instance instance = adminClient.createInstance(createInstanceRequest);
+        System.out.printf("PRODUCTION type instance %s created successfully%n", instance.getId());
+      } catch (AlreadyExistsException e) {
+        System.err.println("Failed to create instance, already exists: " + e.getMessage());
+        System.exit(0);
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: Please re-throw the exception instead

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      System.out.println("The following zones are unavailable: " + e.getUnavailableZones());
+      System.out.println("But the following instances are reachable: " + e.getInstances());
+    }
+    // [END bigtable_list_instances]
+  }
+
+  /**
+   * Demonstrates how to get an instance.
+   *
+   * @return instance
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: Please remove the `@return` its not documenting anything

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    String fakeInstance = generateId();
+    String fakeCluster = generateId();
+    InstanceAdmin testInstanceAdmin =
+        new InstanceAdmin(projectName.getProject(), fakeInstance, fakeCluster);
+    testInstanceAdmin.createProdInstance();
+    assertTrue(adminClient.exists(fakeInstance));
+
+    // Deletes an instance.
+    testInstanceAdmin.deleteInstance();
+    assertTrue(!adminClient.exists(fakeInstance));
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java]: Please use assertFalse

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/** Integration tests for {@link com.google.cloud.examples.bigtable.InstanceAdmin} */
+public class ITInstanceAdmin {
+
+  private static final String INSTANCE_PROPERTY_NAME = "bigtable.instance";
+  private static final String INSTANCE_PREFIX = "instanceadmin";
+  private static final String CLUSTER = "cluster" + System.currentTimeMillis();
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java]: Why can't the cluster id be deterministic?

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  private static String clusterId;
+  private static String instanceId;
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java]: clusterId & instanceId shoud not be static

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+ *   creates production instance
+ *   lists instances
+ *   gets instance
+ *   lists clusters
+ *   adds cluster
+ *   deletes cluster
+ *   deletes instance
+ * </pre>
+ */
+public class InstanceAdmin {
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: Please suffix this with Example or Snippets (I don't know whats the difference)

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  private static BigtableInstanceAdminClient adminClient;
+  private static InstanceAdmin instanceAdmin;
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java]: instanceAdmin is not static

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  private static ProjectName projectName;
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdmin.java]: please separate instance & static variables

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    // [END bigtable_create_cluster]
+  }
+
+  /** Demonstrates how to delete a cluster from an instance. */
+  public void deleteCluster() {
+    System.out.printf("%nDeleting cluster %s from instance: %s%n", CLUSTER, instanceId);
+    // [START bigtable_delete_cluster]
+    try {
+      adminClient.deleteCluster(instanceId, CLUSTER);
+      adminClient.listClusters(instanceId);
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdmin.java]: why is listClusters here?

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdminExample.java [hunk 12]
```diff
@@ -39,17 +40,17 @@
...
 
   private static final String INSTANCE_PROPERTY_NAME = "bigtable.instance";
   private static final String INSTANCE_PREFIX = "instanceadmin";
-  private static final String CLUSTER = "cluster" + System.currentTimeMillis();
-  private static String clusterId;
-  private static String instanceId;
-  private static BigtableInstanceAdminClient adminClient;
-  private static InstanceAdmin instanceAdmin;
+  private static final String CLUSTER = "cluster";
   private static ProjectName projectName;
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdminExample.java]: Please avoid using typesafe names

File: google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdminExample.java [hunk 13]
```diff
@@ -98,14 +99,14 @@ public void testCreateAndDeleteInstance() throws IOException {
     // Creates an instance.
     String fakeInstance = generateId();
     String fakeCluster = generateId();
```
Comment (igorbernstein2) [inline google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdminExample.java]: Why are these fake?
Why can't cluster name be deterministic?

Comment (elisheva-qlogic) [reply google-cloud-examples/src/test/java/com/google/cloud/examples/bigtable/ITInstanceAdminExample.java]: If a cluster with that name exists in a different instance, AlreadyExists exception will be thrown. That't why I used generateId() method here. Is that okay?

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    // Checks if instance exists, creates instance if does not exists.
+    if (!adminClient.exists(instanceId)) {
+      System.out.println("Instance does not exist, creating a PRODUCTION instance");
+      // [START bigtable_create_prod_instance]
+      // Creates a Production Instance with the ID "ssd-instance",
+      // cluster id "ssd-cluster", 3 nodes and location "us-central1-f".
+      CreateInstanceRequest createInstanceRequest =
+          CreateInstanceRequest.of(instanceId)
+              .addCluster(clusterId, "us-central1-f", 3, StorageType.SSD)
+              .setType(Type.PRODUCTION)
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java]: Type is pretty ambiguous, please use Instance.Type.PRODUCTION

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+              .addLabel("example", "instance_admin");
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java]: maybe make the label a bit more meaningful:
`.addLabel("department", "accounting")`

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+    // [START bigtable_list_instances]
+    try {
+      List<Instance> instances = adminClient.listInstances();
+      for (Instance instance : instances) {
+        System.out.println(instance.getId());
+      }
+    } catch (PartialListInstancesException e) {
+      System.out.println("Failed to list instances: " + e.getMessage());
+      System.out.println("The following zones are unavailable: " + e.getUnavailableZones());
+      System.out.println("But the following instances are reachable: " + e.getInstances());
```
Comment (igorbernstein2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigtable/InstanceAdminExample.java]: System.err for error please

**Comment by igorbernstein2** [comment]
thanks!

**Pull request merged**